orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

```diff
@@ -370,143 +392,176 @@ def test_input_requires_grad_backward_creates_input_grad(device):
+        prediction = task_layers(backbone_layers(x))
+        loss = prediction.sum()
+        loss.backward()
 
     for step in range(10):
-        x1 = torch.randn(N, D_in, device='cuda', requires_grad=True)
-        x2 = torch.randn(N, D_in, device='cuda', requires_grad=True)
-        assert x1.grad is None and x2.grad is None
-
-        prediction1 = model1(model(x1))
-        s1 = prediction1.sum()
-        s1.backward()
-
-        prediction2 = model2(model(x2))
-        s2 = prediction2.sum()
-        s2.backward()
-
-        assert x1.grad is not None and x2.grad is not None
-        for param in model.parameters():
-            assert param.grad is not None
-            param.grad = None
-        for param in model1.parameters():
-            assert param.grad is not None
-            param.grad = None
-        for param in model2.parameters():
-            assert param.grad is not None
-            param.grad = None
+        # Run task 1
+        x1 = torch.randn(N, D_in, device=device)
+        run_step(pt_model0, pt_model1, x1)
+        run_step(ort_model0, ort_model1, x1)
+
+        _check_model_gradients_match_and_reset(pt_model0, ort_model0, reset_grad=False)
+        _check_model_gradients_match_and_reset(pt_model1, ort_model1)
+
+        # Run task 2
+        x2 = torch.randn(N, D_in, device=device)
+        run_step(pt_model0, pt_model2, x1)
+        run_step(ort_model0, ort_model2, x1)
+
+        _check_model_gradients_match_and_reset(pt_model0, ort_model0, reset_grad=True)
+        _check_model_gradients_match_and_reset(pt_model2, ort_model2)
 
 def test_multiple_chained_ortmodules_training():
+    device = 'cuda'
     N, D_in, H, D_out = 32, 128, 500, 128
-    model1 = NeuralNetSinglePositionalArgument(D_in, H, D_out).to('cuda')
-    model2 = NeuralNetSinglePositionalArgument(D_in, H, D_out).to('cuda')
-    model1 = ORTModule(model1)
-    model2 = ORTModule(model2)
-
-    all_params = list(model1.parameters()) + list(model2.parameters())
+    pt_model1 = NeuralNetSinglePositionalArgument(D_in, H, D_out).to(device)
+    pt_model2 = NeuralNetSinglePositionalArgument(D_in, H, D_out).to(device)
+    ort_model1 = ORTModule(copy.deepcopy(pt_model1))
+    ort_model2 = ORTModule(copy.deepcopy(pt_model2))
+
+    def run_step(layers1, layers2, x):
+        prediction = layers2(layers1(x))
+        loss = prediction.sum()
+        loss.backward()
 
     for step in range(10):
-        x = torch.randn(N, D_in, device='cuda', requires_grad=True)
-        output1 = model1(x)
-        output2 = model2(output1)
-        s = output2.sum()
-        s.backward()
+        x = torch.randn(N, D_in, device=device, requires_grad=True)
+        run_step(pt_model1, pt_model2, x)
+        run_step(ort_model1, ort_model2, x)
 
-        assert x.grad is not None
-        for param in all_params:
-            assert param.grad is not None
-            param.grad = None
+        _check_model_gradients_match_and_reset(pt_model1, ort_model1)
```

Comment:
When input x is requires_grad, do we need to copy it for out, and also use _assert_tensor_value_matches to check the grad tensors for input x and ort_x?

Reply:
yes, need to clone the input data, if we wish to verify the gradient for inputs. 

see test_multiple_overlapping_forward_backward_calls, I have an example there.